tests/integration/construct_component_methods/testcomponent/index.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright 2016-2021, Pulumi Corporation.  All rights reserved.
+
+import * as pulumi from "@pulumi/pulumi";
+import * as provider from "@pulumi/pulumi/provider";
+
+interface ComponentArgs {
+    first: pulumi.Input<string>;
+    second: pulumi.Input<string>;
+}
+
+class Component extends pulumi.ComponentResource {
+    public readonly first!: pulumi.Output<string>;
+    public readonly second!: pulumi.Output<string>;
+
+    constructor(name: string, args: ComponentArgs, opts?: pulumi.ComponentResourceOptions) {
+        if (opts?.urn) {
+            args = <any>{
+                first: undefined,
+                second: undefined,
+            };
+        }
+
+        super("testcomponent:index:Component", name, args, opts);
+
+        // Skip further initialization when being initialized from a resource reference.
+        if (opts?.urn) {
+            return;
+        }
+
+        this.registerOutputs(args);
+    }
+
+    getMessage(name: pulumi.Input<string>): pulumi.Output<string> {
+        return pulumi.all([this.first, this.second, name]).apply(([frst, scnd, nm]) => `${frst} ${scnd}, ${nm}!`);
+    }
+}
+
+class Provider implements provider.Provider {
+    public readonly version = "0.0.1";
+
+    constructor() {
+        // Register any resources that can come back as resource references that need to be rehydrated.
+        pulumi.runtime.registerResourceModule("testcomponent", "index", {
+            version: this.version,
+            construct: (name, type, urn) => {
+                switch (type) {
+                    case "testcomponent:index:Component":
+                        return new Component(name, <any>undefined, { urn });
+                    default:
+                        throw new Error(`unknown resource type ${type}`);
+                }
+            },
+        });
```

Comment:
Similarly, it's annoying that component authors would have to write this in order to provide support for methods (for `__self__` to come back as a resource reference.

I'm actually wondering whether we could/should implicitly register any resource the first time it is created (in the `Resource` base class), so component authors don't have to even think about or write this.

Reply:
> I'm actually wondering whether we could/should implicitly register any resource the first time it is created (in the Resource base class), so component authors don't have to even think about or write this.

Ah, and this is a problem b/c component authors need these to be registered and don't have the benefit of the schema-generated registrations. I think that in the longer term, we can probably address this with code generation as well, but what you suggest would probably work for now. Given that the registrations would be taking place in the implementing component, I don't think that it would be likely (or perhaps possible) to see a reference to a resource prior to some instance of that resource being created.

Reply:
>Similarly, it's annoying that component authors would have to write this in order to provide support for methods (for __self__ to come back as a resource reference.

Another option is to just require that call authors import their own generated SDKs. I have done this in working with astro.